### PR TITLE
fixed ZipArchive issue in PackageArchiveReader and PluginPackageReader

### PR DESCRIPTION
**Pull request by zhili1208**
this is the second part of https://github.com/NuGet/NuGet.Client/pull/2296

Fixed it in  PackageArchiveReader and PluginPackageReader

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs [hunk 1]
```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
...
+            if(!fullPath.StartsWith(normalizedDestination, StringComparison.Ordinal) || fullPath.Length == normalizedDestination.Length)
+            {
+                throw new UnsafePackageEntryException(string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.ErrorUnsafePackageEntry,
+                    packageIdentity.ToString()));
+            }
+        }
+
+        protected string NormalizesDirectoryPath(string path)
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: rename to `NormalizeDirectoryPath`

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs [hunk 2]
```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
...
+        {
+            if (!path.EndsWith(Path.DirectorySeparatorChar.ToString()))
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: You can consider `path[path.Length-1] != Path.DirectorySeparatorChar`, might be a bit faster than doing string comparison.

**Review by PatoBeltran — APPROVED** [review]

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs [hunk 3]
```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
...
+        /// Eg: file entry like ../../foo.dll can get outside of the expected extraction path.
+        /// </summary>
+        protected static void ValidatePackageEntry(string normalizedDestination, string normalizedFilePath, PackageIdentity packageIdentity)
+        {
+            // Destination and filePath must be normalized.
+            var fullPath = Path.GetFullPath(Path.Combine(normalizedDestination, normalizedFilePath));
+
+            if(!fullPath.StartsWith(normalizedDestination, StringComparison.Ordinal) || fullPath.Length == normalizedDestination.Length)
+            {
+                throw new UnsafePackageEntryException(string.Format(
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs:551]: Where is this caught for `PluginPackageReader` ?

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs [hunk 4]
```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
...
+                    CultureInfo.CurrentCulture,
+                    Strings.ErrorUnsafePackageEntry,
+                    packageIdentity.ToString()));
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: `ToString` is not needed.

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs [hunk 5]
```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
...
+                path += Path.DirectorySeparatorChar;
+            }
+
+            return Path.GetFullPath(path);
+        }
+
+        protected static void ValidatePackageEntries(string normalizedDestination, IEnumerable<string> packageFiles, PackageIdentity packageIdentity)
+        {
+            // Check all package entries.
+            packageFiles.All(p =>
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: Simplify to - 
```
            // Check all package entries.
            packageFiles.ForEach(p => ValidatePackageEntry(normalizedDestination, p, packageIdentity));
```

**Review by mishra14 — APPROVED** [review]

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs (hunk 5 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: ForEach is only for List().

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs (hunk 3 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs:551]: don't need to catch it here, just let it throw

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs (hunk 2 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: EndsWith() actually did almost like that check

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs (hunk 5 above)
Comment (mishra14) [reply src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs]: this works in the code. I have tested it.

File: src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs (hunk 3 above)
Comment (mishra14) [reply src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs:551]: who catches it?

Comment (zhili1208) [reply src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs:551]: the outside caller, it can be UI or PMC or command.

**Pull request merged**